### DISCUSSION TITLE
Use PythonCall inside custom loss function

### DISCUSSION THREAD

**Discussion by joshkamm** [MilesCranmer/PySR#647] [Q&A]
I'm working with @SoumiKDaS1701 trying to customize PySR to constrain aspects of the generated functions like the asymptotic limit or its derivative at long range. In order to do this, we're trying to call sympy inside of a custom loss function using PythonCall which seems like the inverse of how PySR calls into Julia. Here's our code, which looks similar to [this example](https://github.com/MilesCranmer/PySR/discussions/617).

```python
jl.seval(
    """
    import Pkg
    Pkg.add("DynamicExpressions")

    using PythonCall
    using DynamicExpressions: string_tree

    function my_custom_objective(tree, dataset::Dataset{T,L}, options)::L where {T,L}
       
       prediction, flag = eval_tree_array(tree, dataset.X, options)
       if !flag
            return L(Inf)
       end
       
       eq_str = string_tree(tree, options)

       # import and call our python code that uses sympy
       utils = pyimport("cheml_pysr.utils")
       limit1 = utils.eval_function_limit(x_val=0, f=eq_str)
       limit2 = utils.eval_derivative_limit(x_val=0, f=eq_str)

       lim_loss = abs(0.5 - pyconvert(Float32, limit1))
       deriv_loss = abs(0.25 - pyconvert(Float32, limit2))

       prediction_loss = sum((prediction .- dataset.y) .^ 2) / dataset.n

       lambda1 = 10
       lambda2 = 10
       loss = prediction_loss + lambda1*lim_loss + lambda2*deriv_loss

       return loss
    end
    """
)
```

Both of us are new to Julia and the interfacing between Python and Julia so we're having some difficulty getting this to work.

Our immediate issue is getting a segmentation fault which might be related to [this issue](https://github.com/JuliaPy/PythonCall.jl/issues/219)

```
[26897] signal (11.1): Segmentation fault
in expression starting at none:0
unknown function (ip: 0x50b07b)
PyUnicode_DecodeUTF8 at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/C/pointers.jl:297 [inlined]
pystr_fromUTF8 at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:574 [inlined]
pystr_fromUTF8 at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:575 [inlined]
pystr at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:583 [inlined]
Py at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/Core/Py.jl:139 [inlined]
macro expansion at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/Core/Py.jl:131 [inlined]
pyimport at /export/zimmerman/soumikd/.julia/packages/PythonCall/S5MOg/src/Core/builtins.jl:1444
my_custom_objective at ./none:15
evaluator at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/LossFunctions.jl:92
unknown function (ip: 0x7feb8fb8f4b0)

_jl_invoke at /cache/build/builder-amdci4-2/julialang/julia-release-1-dot-10/src/gf.c:2895 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-2/julialang/julia-release-1-dot-10/src/gf.c:3077
#eval_loss#3 at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/LossFunctions.jl:108
eval_loss at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/LossFunctions.jl:97 [inlined]
#score_func#5 at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/LossFunctions.jl:164 [inlined]
score_func at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/LossFunctions.jl:161 [inlined]
#PopMember#2 at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/PopMember.jl:99
PopMember at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/PopMember.jl:88 [inlined]
PopMember at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/PopMember.jl:88 [inlined]
#2 at ./none:0 [inlined]
iterate at ./generator.jl:47 [inlined]
collect at ./array.jl:834
#Population#1 at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/Population.jl:49 [inlined]
Population at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/Population.jl:35 [inlined]
macro expansion at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/SymbolicRegression.jl:765 [inlined]
#54 at /export/zimmerman/soumikd/.julia/packages/SymbolicRegression/FtJSD/src/SearchUtils.jl:116
unknown function (ip: 0x7feb8fb99147)
_jl_invoke at /cache/build/builder-amdci4-2/julialang/julia-release-1-dot-10/src/gf.c:2895 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-2/julialang/julia-release-1-dot-10/src/gf.c:3077
jl_apply at /cache/build/builder-amdci4-2/julialang/julia-release-1-dot-10/src/julia.h:1982 [inlined]
start_task at /cache/build/builder-amdci4-2/julialang/julia-release-1-dot-10/src/task.c:1238
Allocations: 22890647 (Pool: 22860626; Big: 30021); GC: 29
Segmentation fault (core dumped)
```

We would appreciate some guidance either on this specific issue or more broadly on how we would go about constraining the functions generated by PySR based on analytical properties.

**Comment by MilesCranmer** [comment]
Does it work if you turn parallelism off? i.e., `numprocs=0, multithreading=false`. (Python doesn't support multithreading, so if you call back into Python, you likely need to avoid parallelism, so the code doesn't violate the GIL https://wiki.python.org/moin/GlobalInterpreterLock)

**Reply by joshkamm** [reply]
We had tried `multithreading=false` and `procs=1` because we assumed 1 meant single threaded. We just tried `procs=0` (does this set numprocs corrently internally?) and it looks like PySR is getting further than before but we're still getting a segmentation fault after it has already gone through a number of iterations but before it prints the hall of fame.

```
[13845] signal (11.1): Segmentation fault
in expression starting at none:0
unknown function (ip: 0x50b07b)
Segmentation fault (core dumped)
```

Here is our PySRRegressor object...
```python
model = PySRRegressor(
    niterations=10,  # < Increase me for better results
    populations= 10,
    binary_operators=["+", "*", "/", "-"],
    unary_operators=[
        "exp",
        # "inv(x) = 1/x",
        # "square"
        # "sqrt"
    ],
    maxsize=7,
    nested_constraints={"exp": {"exp":0}},
    bumper= False,
    extra_sympy_mappings={"inv": lambda x: 1 / x},
    procs=0,
    multithreading=False,
    # ^ Define operator for SymPy as well
    # elementwise_loss="loss(prediction, target) = (prediction - target)^2",
    # ^ Custom loss function (julia syntax)
    loss_function= "my_custom_objective",
    equation_file="/export/zimmerman/soumikd/symbolic_regression/cheml-pysr/tests/test_data/test.csv"
)
```

**Reply by MilesCranmer** [reply]
Hm. Maybe try turning off the following line in your local copy of PySR? https://github.com/MilesCranmer/PySR/blob/476f57325e01b3aa2160d1a046106ed5b7236dde/pysr/sr.py#L1746

This line turns off PythonCall garbage collection – required for doing any multithreading stuff in Julia, but perhaps it causes memory issues if you are trying to call back into Python? I'm not really sure.

**Reply by MilesCranmer** [reply]
By the way, I should say I would try to avoid Python if you can. The speed of the loss function is quite important, and calling Python every single time is going to be super super slow. So in addition to the memory issues you are seeing, it might also help with speed.

Unless you **HAVE** to call Python. But if you don't, I would try to translate your loss function to Julia if possible. 

A lot of the stuff you can probably just do directly on the equation type using the [utility functions](https://symbolicml.org/DynamicExpressions.jl/dev/utils/). Make a new discussion thread if you want some help. Converting to SymPy every step is a bit overkill and will result in like a 10,000x slowdown.

**Reply by joshkamm** [reply]
Thank you for your feedback. We're considering both numerically computed asymptotic constraints like mentioned [here](https://github.com/MilesCranmer/PySR/discussions/324) as well as symbolic asymptotic constraints. We figured there would be a penalty in cost per evaluation for symbolic evaluation but that more accurate feedback might reduce the number of iterations necessary so we planned to benchmark both. But we didn't realize the cost would be so dramatic. If you're confident that evaluating the loss function would become a prohibitively expensive bottleneck if we call back into python / sympy, we could try to stick to what can be quickly evaluated in Julia. I just found [this pull request](https://github.com/MilesCranmer/PySR/pull/276) that looks like it has similar goals to some of the things we're trying to do.

**Reply by MilesCranmer** [reply]
> If you're confident that evaluating the loss function would become a prohibitively expensive bottleneck if we call back into python / sympy

I am pretty confident that this is true. calling into python within the loss function will be a massive bottleneck. Especially if it is SymPy code which is mostly pure-Python based (rather than a Python library with a compiled backend like numpy). Better to operate directly on the equation datatype (just post another thread describing what you want to do and maybe I or others can help).

Worst case you could try SymbolicUtils.jl. It will be a bit slower than sticking with the Node type but it’s more SymPy-like.
